core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/environment/network/SubnetSelector.java

```diff
@@ -17,21 +22,42 @@
 public class SubnetSelector {
     private static final Logger LOGGER = LoggerFactory.getLogger(SubnetSelector.class);
 
-    public Optional<CloudSubnet> chooseSubnet(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone, boolean allowRandom) {
-        Optional<CloudSubnet> cloudSubnet;
+    @Inject
+    private EntitlementService entitlementService;
+
+    public Optional<CloudSubnet> chooseSubnet(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone,
+            boolean allowRandom) {
+        return chooseSubnet(preferredSubnetId, subnetMetas, availabilityZone, allowRandom, false);
+    }
+
+    @VisibleForTesting
+    Optional<CloudSubnet> chooseSubnetPreferPublic(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone) {
+        return chooseSubnet(preferredSubnetId, subnetMetas, availabilityZone, false, true);
+    }
+
+    private Optional<CloudSubnet> chooseSubnet(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone,
+            boolean allowRandom, boolean preferPublic) {
+        Optional<CloudSubnet> cloudSubnet = Optional.empty();
         if (StringUtils.isNotEmpty(preferredSubnetId)) {
             cloudSubnet = findSubnetById(subnetMetas, preferredSubnetId);
         } else if (StringUtils.isNotEmpty(availabilityZone)) {
             LOGGER.debug("Choosing subnet by availability zone {}", availabilityZone);
-            cloudSubnet = subnetMetas.values().stream()
-                .filter(s -> StringUtils.isNotEmpty(s.getAvailabilityZone()) &&
-                    s.getAvailabilityZone().equals(availabilityZone))
-                .findFirst();
+            if (preferPublic) {
```

Comment:
Can we have debug logs here that indicates which path selected the the subnet based on what flags?